File: tobas_core/tobas_rqt_tools/src/tobas_rqt_tools/widgets/vertical_tab_widget.py
from PyQt5.QtCore import Qt, QRect, QPoint
from PyQt5.QtWidgets import QTabBar, QTabWidget, QStyle, QStylePainter, QStyleOptionTab


class _TabBar(QTabBar):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

    def tabSizeHint(self, index):
        s = QTabBar.tabSizeHint(self, index)
        if s.width() < s.height():
            s.transpose()
        s.scale(s.width() * 2, s.height() * 2, Qt.KeepAspectRatio)
        return s

    def paintEvent(self, _):
        painter = QStylePainter(self)
        style_option = QStyleOptionTab()

        for i in range(self.count()):
            self.initStyleOption(style_option, i)
            painter.drawControl(QStyle.CE_TabBarTabShape, style_option)
            painter.save()

            s = style_option.rect.size()
            s.scale(s.width() * 2, s.height() * 2, Qt.KeepAspectRatio)
            rect = QRect(QPoint(), s)
            rect.moveCenter(style_option.rect.center())
            style_option.rect = rect

            center = self.tabRect(i).center()
            painter.translate(center)
            painter.rotate(90)
            painter.translate(center * -1)
            painter.drawControl(QStyle.CE_TabBarTabLabel, style_option)
            painter.restore()


class VerticalTabWidget(QTabWidget):
    def __init__(self, parent=None):
        super().__init__(parent=parent)
        self.setTabBar(_TabBar(self))
        self.setTabPosition(QTabWidget.West)

File: tobas_gui/tobas_qt_tools/src/widgets/vertical_tab_widget.cpp
#include <QTabBar>
#include <QWheelEvent>
#include <QStylePainter>
#include <QStyleOptionTab>

#include "tobas_qt_tools/widgets/vertical_tab_widget.hpp"

namespace qt
{
QSize VerticalTabBar::tabSizeHint(int index) const
{
  auto s = super::tabSizeHint(index);
  if (s.width() < s.height())
    s.transpose();
  s.scale(s.width() * 2, s.height() * 2, Qt::KeepAspectRatio);
  return s;
}

void VerticalTabBar::paintEvent(QPaintEvent*)
{
  QStylePainter painter(this);
  QStyleOptionTab style_option;

  for (int i = 0; i < count(); ++i)
  {
    initStyleOption(&style_option, i);
    painter.drawControl(QStyle::CE_TabBarTabShape, style_option);
    painter.save();

    auto s = style_option.rect.size();
    s.scale(s.width() * 2, s.height() * 2, Qt::KeepAspectRatio);
    QRect rect(QPoint(), s);
    rect.moveCenter(style_option.rect.center());
    style_option.rect = rect;

    const auto center = tabRect(i).center();
    painter.translate(center);
    painter.rotate(90);
    painter.translate(center * -1);
    painter.drawControl(QStyle::CE_TabBarTabLabel, style_option);
    painter.restore();
  }
}

VerticalTabWidget::VerticalTabWidget(QWidget* parent) : super(parent)
{
  setTabBar(new VerticalTabBar());
  setTabPosition(QTabWidget::West);
}
}  // namespace qt
